### PR TITLE
Smc update succ gen

### PR DESCRIPTION
**Pull request by MikkelTJensen**

### CODE REVIEW

File: include/PetriEngine/PetriNet.h [hunk 1]
```diff
@@ -71,6 +71,9 @@ namespace PetriEngine {
         uint32_t numberOfTransitions() const {
             return _ntransitions;
         }
+        const std::vector<uint32_t> placeToPtrs() const{
+            return _placeToPtrs;
+        }
```
Comment (t-bsh) [inline include/PetriEngine/PetriNet.h:76]: no touch bad

File: include/SMC/SMCMain.h [hunk 2]
```diff
@@ -18,11 +18,12 @@
...
 #include "PetriEngine/PetriNet.h"
+#include "PetriEngine/PQL/PQL.h"
 #include "PetriEngine/options.h"
 
 namespace SMC{
     double SMCMain(const PetriEngine::PetriNet *net,
-                         options_t &options);
+                         options_t &options,
+                         const PetriEngine::PQL::Condition_ptr &query);
 }
```
Comment (t-bsh) [inline include/SMC/SMCMain.h]:
```suggestion
}

```

File: include/PetriEngine/PetriNet.h (hunk 1 above)
Comment (faje18) [reply include/PetriEngine/PetriNet.h:76]: ?? what happen here ??

File: include/SMC/SuccessorGeneration/SMCSuccessorGenerator.h [hunk 3]
```diff
@@ -27,9 +27,15 @@ namespace SMC{
...
+        //     return has_suc;
+        // }
+
+        bool next(Structures::State &write, uint32_t &tindex);
+
+        void fire(Structures::State &write, uint32_t tindex){
+            _fire(write, tindex);
         }
     };
 }
```
Comment (t-bsh) [inline include/SMC/SuccessorGeneration/SMCSuccessorGenerator.h]:
```suggestion
}

```

File: src/SMC/SMCMain.cpp [hunk 4]
```diff
@@ -18,77 +18,58 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+//#include <stdlib.h>
```
Comment (t-bsh) [inline src/SMC/SMCMain.cpp]:
```suggestion
```

File: include/PetriEngine/PetriNet.h (hunk 1 above)
Comment (fulei345) [reply include/PetriEngine/PetriNet.h:76]:
```suggestion
```

File: include/SMC/SuccessorGeneration/SMCSuccessorGenerator.h [hunk 5]
```diff
@@ -27,9 +27,15 @@ namespace SMC{
         public:
         SMCSuccessorGenerator(const PetriNet &net);
     
-        bool next(Structures::State &write, uint32_t &tindex) {
-            bool has_suc = SuccessorGenerator::next(write, tindex);
-            return has_suc;
+        // bool next(Structures::State &write, uint32_t &tindex) {
+        //     bool has_suc = next(write, tindex);
+        //     return has_suc;
+        // }
```
Comment (fulei345) [inline include/SMC/SuccessorGeneration/SMCSuccessorGenerator.h]:
```suggestion
```

**Review by fulei345 — CHANGES_REQUESTED** [review]

File: src/SMC/SMCMain.cpp [hunk 6]
```diff
@@ -18,77 +18,58 @@
...
-                initial_size += qm0[i];
+namespace SMC
+{
+    bool SMCRun(SMCSuccessorGenerator &sgen,
+                const PetriNet *net,
+                const PQL::Condition_ptr &query,
+                int max_depth)
+    {
+        int current_depth = 0;
+        uint32_t tindex = 0;
```
Comment (t-bsh) [inline src/SMC/SMCMain.cpp:40]:
```suggestion
        uint32_t tindex = 0;
        
```

File: src/SMC/SMCMain.cpp [hunk 7]
```diff
@@ -18,77 +18,58 @@
...
+        Structures::State write(net->makeInitialMarking());
+        sgen.prepare(&write);
+        sgen.reset();
```
Comment (t-bsh) [inline src/SMC/SMCMain.cpp:44]:
```suggestion
        sgen.reset();
        
```

File: src/SMC/SMCMain.cpp [hunk 8]
```diff
@@ -18,77 +18,58 @@
...
-                // }
-                std::cout << "\nSMC Simulation:" << std::endl;
-                std::cout << "Simulate " << options.smcruns << " runs, with " << options.smcdepth << " max depth." << std::endl;
+    double SMCMain(const PetriNet *net,
+                         options_t &options,
+                         const PQL::Condition_ptr &query)
+    {
+        int total_runs = 0;
+        int successful_runs = 0;
+        SMCSuccessorGenerator sgen(*net);
```
Comment (t-bsh) [inline src/SMC/SMCMain.cpp:65]:
```suggestion
        SMCSuccessorGenerator sgen(*net);
        
```

File: src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp [hunk 9]
```diff
@@ -21,9 +21,52 @@
...
 
     SMCSuccessorGenerator::SMCSuccessorGenerator(const PetriNet &net)
     : SuccessorGenerator(net){}
+
+    bool SMCSuccessorGenerator::next(Structures::State& write, uint32_t &tindex)
+    {
+        _parent = &write;
+        u_int32_t tcurrent = std::numeric_limits<uint32_t>::max();
+        int n = 0;
+        for (_suc_pcounter = 0; _suc_pcounter < _net.numberOfPlaces(); ++_suc_pcounter)
```
Comment (t-bsh) [inline src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp]:
```suggestion
        for (; _suc_pcounter < _net.numberOfPlaces(); ++_suc_pcounter)
```

File: src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp [hunk 10]
```diff
@@ -21,9 +21,52 @@
...
     using namespace PetriEngine;
 
     SMCSuccessorGenerator::SMCSuccessorGenerator(const PetriNet &net)
     : SuccessorGenerator(net){}
+
+    bool SMCSuccessorGenerator::next(Structures::State& write, uint32_t &tindex)
+    {
+        _parent = &write;
+        u_int32_t tcurrent = std::numeric_limits<uint32_t>::max();
+        int n = 0;
```
Comment (t-bsh) [inline src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp:35]:
```suggestion
        int n = 0;
        
```

File: src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp [hunk 11]
```diff
@@ -21,9 +21,52 @@
...
+                for (; tindex < last; ++tindex)
+                {
+                    if (!checkPreset(tindex))
+                    {
+                        continue;
+                    }
+                    else
+                    {
+                        // In here we check the tindex against our currently chosen transition
+                        // TODO increment n with potency instead
```
Comment (t-bsh) [inline src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp]:
```suggestion
                        // TODO non-uniform, increment n with potency instead n+=m
```

File: src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp [hunk 12]
```diff
@@ -21,9 +21,52 @@
...
+                        ++n;
+                        double randomNum = (double)rand()/RAND_MAX;
```
Comment (t-bsh) [inline src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp:54]:
```suggestion
                        double randomNum = (double)rand()/RAND_MAX;
                        // TODO non-uniform, (double)m/(double)n
```

File: src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp [hunk 13]
```diff
@@ -21,9 +21,52 @@
...
+        // Set tindex to chosen transition so we can read it in main
+        tindex = tcurrent;
+        if(tcurrent != std::numeric_limits<uint32_t>::max())
+        {
+            _fire(write, tcurrent);
+            return true;
+        }
+        return false;
+    }
 }
```
Comment (t-bsh) [inline src/SMC/SuccessorGeneration/SMCSuccessorGenerator.cpp]:
```suggestion
}

```

File: src/main.cpp [hunk 14]
```diff
@@ -272,6 +261,18 @@ int main(int argc, const char** argv) {
 
         options.queryReductionTimeout = 0;
 
+        //-------------------------- SMC Simulation --------------------------//
+        if (options.smc){
+            PetriNetBuilder b2(builder);
```
Comment (t-bsh) [inline src/main.cpp]:
```suggestion
```

**Review by fulei345 — APPROVED** [review]

**Review by t-bsh — APPROVED** [review]
:)

**Review by faje18 — APPROVED** [review]

**Pull request merged**